### PR TITLE
certificate validation on client side; check stored fingerprint on server start-up; non-optional fingerprint parsing

### PR DESCRIPTION
**Pull request by efim-poberezkin**

### CODE REVIEW

File: apps/smp-server/Main.hs [hunk 1]
```diff
@@ -113,7 +114,8 @@ makeConfig IniOpts {serverPort, enableWebsockets, serverPrivateKeyFile, serverCe
 
 printConfig :: ServerConfig -> IO ()
 printConfig ServerConfig {storeLog} = do
-  -- TODO print certificate hash
+  certificateHash <- readFile certificateHashFile
+  putStr $ "certificate hash: " <> certificateHash
```
Comment (epoberezkin) [inline apps/smp-server/Main.hs:118]: I think we should validate that this hash matches the certificate on start up

File: migrations/20211220_certificate_hash.sql [hunk 2]
```diff
@@ -0,0 +1 @@
+ALTER TABLE servers RENAME COLUMN key_hash TO certificate_hash;
```
Comment (epoberezkin) [inline migrations/20211220_certificate_hash.sql:1]: do we really need it? something makes me worry here...

File: src/Simplex/Messaging/Transport.hs [hunk 3]
```diff
@@ -134,11 +141,14 @@ data ATransport = forall c. Transport c => ATransport (TProxy c)
...
 
+-- | Party name to print on exception handling - for debugging.
+newtype PartyAlias = PartyAlias {partyAlias :: String} deriving (Show)
+
 -- | Run transport server (plain TCP or WebSockets) on passed TCP port and signal when server started and stopped via passed TMVar.
 --
 -- All accepted connections are passed to the passed function.
-runTransportServer :: forall c m. (Transport c, MonadUnliftIO m) => TMVar Bool -> ServiceName -> T.ServerParams -> (c -> m ()) -> m ()
-runTransportServer started port serverParams server = do
+runTransportServer :: forall c m. (Transport c, MonadUnliftIO m) => PartyAlias -> TMVar Bool -> ServiceName -> T.ServerParams -> (c -> m ()) -> m ()
```
Comment (efim-poberezkin) [inline src/Simplex/Messaging/Transport.hs:150]: @epoberezkin could be just String [instead of PartyAlias] but I thought it's better to give it a name since it's part of API, do you see drawbacks?

Also it can be removed altogether but then it's hard to figure out what exactly fails on tests

File: src/Simplex/Messaging/Transport.hs [hunk 4]
```diff
@@ -223,6 +234,61 @@ loadTLSServerParams certificateFile privateKeyFile =
...
+      T.clientHooks = def {T.onServerCertificate = \_ _ _ -> validateCertificateChain certificateHash host p},
+      T.clientSupported = supportedParameters
+    }
+
+validateCertificateChain :: C.CertificateHash -> HostName -> ByteString -> X.CertificateChain -> IO [XV.FailedReason]
+validateCertificateChain _ _ _ (X.CertificateChain []) = pure [XV.EmptyChain]
+validateCertificateChain expectedHash host port cc@(X.CertificateChain sc@[cert]) = do
+  let certificateHash = C.certificateHash . X.getCertificate $ cert
+      fr = checkHash certificateHash
+  -- putStrLn $ "certificate hash: " <> (B.unpack . encode . C.unCertificateHash) certificateHash
```
Comment (efim-poberezkin) [inline src/Simplex/Messaging/Transport.hs:251]: kept these until I fix that test

File: src/Simplex/Messaging/Agent.hs [hunk 5]
```diff
@@ -85,7 +85,7 @@ import Simplex.Messaging.Client (SMPServerTransmission)
 import qualified Simplex.Messaging.Crypto as C
 import Simplex.Messaging.Protocol (MsgBody, SndPublicVerifyKey)
 import qualified Simplex.Messaging.Protocol as SMP
-import Simplex.Messaging.Transport (ATransport (..), TProxy, Transport (..), currentSMPVersionStr, loadTLSServerParams, runTransportServer)
+import Simplex.Messaging.Transport
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent.hs:88]: what happened to whitelist everything :)? I've started making them multiline in some cases, but sometimes just import all :) No need to change, please.

File: src/Simplex/Messaging/Agent.hs [hunk 6]
```diff
@@ -112,7 +112,7 @@ runSMPAgentBlocking (ATransport t) started cfg@AgentConfig {tcpPort, agentCertif
     smpAgent _ = do
       -- tlsServerParams not in env to avoid breaking functional api w/t key and certificate generation
       tlsServerParams <- liftIO $ loadTLSServerParams agentCertificateFile agentPrivateKeyFile
-      runTransportServer started tcpPort tlsServerParams $ \(h :: c) -> do
+      runTransportServer (PartyAlias "SMP agent") started tcpPort tlsServerParams $ \(h :: c) -> do
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent.hs:115]: newtype seems unnecessary tbh

File: src/Simplex/Messaging/Agent/Protocol.hs [hunk 7]
```diff
@@ -472,7 +472,7 @@ connModeT = \case
 data SMPServer = SMPServer
   { host :: HostName,
     port :: Maybe ServiceName,
-    keyHash :: Maybe C.KeyHash
+    certificateHash :: C.CertificateHash
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent/Protocol.hs:475]: thinking we should keep maybe (and required in parsing), but next PR

File: src/Simplex/Messaging/Agent/Store/SQLite.hs [hunk 8]
```diff
@@ -603,18 +603,18 @@ instance (ToField a, ToField b, ToField c, ToField d, ToField e, ToField f,
...
 
 upsertServer_ :: DB.Connection -> SMPServer -> IO ()
-upsertServer_ dbConn SMPServer {host, port, keyHash} = do
+upsertServer_ dbConn SMPServer {host, port, certificateHash} = do
   let port_ = serializePort_ port
   DB.executeNamed
     dbConn
     [sql|
-      INSERT INTO servers (host, port, key_hash) VALUES (:host,:port,:key_hash)
+      INSERT INTO servers (host, port, certificate_hash) VALUES (:host,:port,:certificate_hash)
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent/Store/SQLite.hs:611]: what annoys me most is the size, rather than renaming... why not cert_hash?

File: src/Simplex/Messaging/Client.hs [hunk 9]
```diff
@@ -164,7 +164,7 @@ getSMPClient smpServer cfg@SMPClientConfig {qSize, tcpTimeout, smpPing, smpBlock
       thVar <- newEmptyTMVarIO
       action <-
         async $
-          runTransportClient (host smpServer) port' (client t c thVar)
+          runTransportClient (PartyAlias "SMP client") (host smpServer) port' (certificateHash smpServer) (client t c thVar)
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Client.hs:167]: same - no newtype needed

File: src/Simplex/Messaging/Crypto.hs [hunk 10]
```diff
@@ -812,19 +812,19 @@ newtype Key = Key {unKey :: ByteString}
 -- | IV bytes newtype.
 newtype IV = IV {unIV :: ByteString}
 
--- | Key hash newtype.
-newtype KeyHash = KeyHash {unKeyHash :: ByteString} deriving (Eq, Ord, Show)
+-- | Certificate hash newtype.
+newtype CertificateHash = CertificateHash {unCertificateHash :: ByteString} deriving (Eq, Ord, Show)
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Crypto.hs:816]: some yak shaving... let's please not rename in cases like this in the future.

File: src/Simplex/Messaging/Transport.hs (hunk 3 above)
Comment (epoberezkin) [reply src/Simplex/Messaging/Transport.hs:150]: If you want name in the API, it could be just type synonym - this is better:

```
type PartyAlias = String -- or ByteString, still don't see where it is used
```

File: src/Simplex/Messaging/Transport.hs [hunk 11]
```diff
@@ -177,13 +187,14 @@ startTCPServer started port = withSocketsDo $ resolve >>= open >>= setStarted
...
+runTransportClient :: Transport c => MonadUnliftIO m => PartyAlias -> HostName -> ServiceName -> C.CertificateHash -> (c -> m a) -> m a
+runTransportClient partyAlias host port certificateHash client = do
+  let clientParams = mkTLSClientParams host port certificateHash
+  c <- liftIO $ startTCPClient partyAlias host port clientParams
   client c `E.finally` liftIO (closeConnection c)
 
-startTCPClient :: forall c. Transport c => HostName -> ServiceName -> IO c
-startTCPClient host port = withSocketsDo $ resolve >>= tryOpen err
+startTCPClient :: forall c. Transport c => PartyAlias -> HostName -> ServiceName -> T.ClientParams -> IO c
+startTCPClient PartyAlias {partyAlias} host port clientParams = withSocketsDo $ resolve >>= tryOpen err
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:197]: this feels completely unnecessary even as parameter as it can only be "client"

Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:197]:
```suggestion
startTCPClient :: forall c. Transport c => HostName -> ServiceName -> T.ClientParams -> IO c
startTCPClient host port clientParams = withSocketsDo $ resolve >>= tryOpen err
```

File: src/Simplex/Messaging/Transport.hs [hunk 12]
```diff
@@ -202,7 +213,7 @@ startTCPClient host port = withSocketsDo $ resolve >>= tryOpen err
     open addr = do
       sock <- socket (addrFamily addr) (addrSocketType addr) (addrProtocol addr)
       connect sock $ addrAddress addr
-      ctx <- connectTLS "client" clientParams sock
+      ctx <- connectTLS partyAlias clientParams sock
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:216]:
```suggestion
      ctx <- connectTLS "client" clientParams sock
```

File: src/Simplex/Messaging/Transport.hs [hunk 13]
```diff
@@ -134,11 +141,14 @@ data ATransport = forall c. Transport c => ATransport (TProxy c)
...
+-- | Party name to print on exception handling - for debugging.
+newtype PartyAlias = PartyAlias {partyAlias :: String} deriving (Show)
+
 -- | Run transport server (plain TCP or WebSockets) on passed TCP port and signal when server started and stopped via passed TMVar.
 --
 -- All accepted connections are passed to the passed function.
-runTransportServer :: forall c m. (Transport c, MonadUnliftIO m) => TMVar Bool -> ServiceName -> T.ServerParams -> (c -> m ()) -> m ()
-runTransportServer started port serverParams server = do
+runTransportServer :: forall c m. (Transport c, MonadUnliftIO m) => PartyAlias -> TMVar Bool -> ServiceName -> T.ServerParams -> (c -> m ()) -> m ()
+runTransportServer PartyAlias {partyAlias} started port serverParams server = do
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:151]: and now that I see where it is used it seems completely unnecessary too - please remove.
Let's not spend time on such things, it's yak shaving.

Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:151]:
```suggestion
runTransportServer :: forall c m. (Transport c, MonadUnliftIO m) => TMVar Bool -> ServiceName -> T.ServerParams -> (c -> m ()) -> m ()
runTransportServer started port serverParams server = do
```

File: src/Simplex/Messaging/Transport.hs [hunk 14]
```diff
@@ -158,7 +168,7 @@ runTransportServer started port serverParams server = do
     acceptConnection :: Socket -> IO c
     acceptConnection sock = do
       (newSock, _) <- accept sock
-      ctx <- connectTLS "server" serverParams newSock
+      ctx <- connectTLS partyAlias serverParams newSock
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:171]:
```suggestion
      ctx <- connectTLS "server" serverParams newSock
```

File: src/Simplex/Messaging/Crypto.hs (hunk 10 above)
Comment (epoberezkin) [reply src/Simplex/Messaging/Crypto.hs:816]: I actually think let's please rename back

File: src/Simplex/Messaging/Client.hs (hunk 9 above)
Comment (epoberezkin) [inline src/Simplex/Messaging/Client.hs:167]:
```suggestion
          runTransportClient (host smpServer) port' (certificateHash smpServer) (client t c thVar)
```

File: src/Simplex/Messaging/Agent.hs (hunk 6 above)
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent.hs:115]:
```suggestion
      runTransportServer started tcpPort tlsServerParams $ \(h :: c) -> do
```

File: src/Simplex/Messaging/Transport.hs [hunk 15]
```diff
@@ -223,6 +234,61 @@ loadTLSServerParams certificateFile privateKeyFile =
...
+        hooks = XV.defaultHooks
+        checks :: XV.ValidationChecks
+        checks = XV.defaultChecks {XV.checkLeafV3 = False} -- TODO create v3 certificates? https://stackoverflow.com/a/18242720
+        certStore :: XS.CertificateStore
+        certStore = XS.makeCertificateStore sc
+        cache :: XV.ValidationCache
+        cache = XV.exceptionValidationCache [] -- we don't store certificate Fingerprints on clients
+        serviceID :: XV.ServiceID
+        serviceID = (host, port)
+validateCertificateChain _ _ _ (X.CertificateChain (_ : _ : _)) = pure [XV.AuthorityTooDeep]
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:274]:
```suggestion
validateCertificateChain _ _ _ (X.CertificateChain (_ : _ )) = pure [XV.AuthorityTooDeep]
```

I think this is enough here

Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:274]:
```suggestion
validateCertificateChain _ _ _ (X.CertificateChain (_ : _)) = pure [XV.AuthorityTooDeep]
```

I think this is enough here

File: src/Simplex/Messaging/Transport.hs [hunk 16]
```diff
@@ -223,6 +234,61 @@ loadTLSServerParams certificateFile privateKeyFile =
...
+    else pure fr
+  where
+    checkHash :: C.CertificateHash -> [XV.FailedReason]
+    checkHash hash
+      | hash == expectedHash = []
+      | otherwise = [XV.UnknownCA]
+    x509validate :: IO [XV.FailedReason]
+    x509validate = XV.validate X.HashSHA256 hooks checks certStore cache serviceID cc
+      where
+        hooks :: XV.ValidationHooks
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:264]: no need to define types on trivial things in where - it only creates noise

Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:264]:
```suggestion
```

File: src/Simplex/Messaging/Transport.hs [hunk 17]
```diff
@@ -223,6 +234,61 @@ loadTLSServerParams certificateFile privateKeyFile =
...
+        hooks = XV.defaultHooks
+        checks :: XV.ValidationChecks
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:266]:
```suggestion
```

File: src/Simplex/Messaging/Transport.hs [hunk 18]
```diff
@@ -223,6 +234,61 @@ loadTLSServerParams certificateFile privateKeyFile =
...
+        checks = XV.defaultChecks {XV.checkLeafV3 = False} -- TODO create v3 certificates? https://stackoverflow.com/a/18242720
+        certStore :: XS.CertificateStore
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:268]:
```suggestion
```

File: src/Simplex/Messaging/Transport.hs [hunk 19]
```diff
@@ -223,6 +234,61 @@ loadTLSServerParams certificateFile privateKeyFile =
...
+      T.supportedSecureRenegotiation = False,
+      T.supportedGroups = [T.X448, T.X25519]
+    }
+
+getCertificateHash :: FilePath -> IO ByteString
+getCertificateHash certificateFile = do
+  x509 <- SX.readSignedObject certificateFile
+  let certificate = X.getCertificate (head x509) -- we should have only one certificate
+  pure $ (encode . C.unCertificateHash) (C.certificateHash certificate)
+
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:291]: please don't move things around to create unnecessary diff - the location of code in the file is not important, no need to adhere to the sections anywhere other than in module exports - functions can be anywhere.

Please return them where they were.

File: src/Simplex/Messaging/Crypto.hs (hunk 10 above)
Comment (epoberezkin) [reply src/Simplex/Messaging/Crypto.hs:816]: all that was needed to write in comment that it is actually a hash of the whole certificate, not just the key and keep the same, shorter type name

File: src/Simplex/Messaging/Transport.hs [hunk 20]
```diff
@@ -223,6 +234,58 @@ loadTLSServerParams certificateFile privateKeyFile =
...
+  -- putStrLn $ "certificate hash: " <> (B.unpack . encode . C.unCertificateHash) certificateHash
+  -- putStrLn $ "expected hash: " <> (B.unpack . encode . C.unCertificateHash) expectedHash
+  if null fr
+    then x509validate
+    else pure fr
+  where
+    checkHash :: C.CertificateHash -> [XV.FailedReason]
+    checkHash hash
+      | hash == expectedHash = []
+      | otherwise = [XV.UnknownCA]
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:260]:
```suggestion
  -- putStrLn $ "certificate hash: " <> (B.unpack . encode . C.unCertificateHash) certificateHash
  -- putStrLn $ "expected hash: " <> (B.unpack . encode . C.unCertificateHash) expectedHash
  if hash == expectedHash
    then x509validate
    else pure [XV.UnknownCA]
  where
```

File: src/Simplex/Messaging/Transport.hs [hunk 21]
```diff
@@ -223,6 +234,53 @@ loadTLSServerParams certificateFile privateKeyFile =
...
+    then x509validate
+    else pure [XV.UnknownCA]
+  where
+    x509validate :: IO [XV.FailedReason]
+    x509validate = XV.validate X.HashSHA256 hooks checks certStore cache serviceID cc
+      where
+        hooks = XV.defaultHooks
+        checks = XV.defaultChecks {XV.checkLeafV3 = False} -- TODO create v3 certificates? https://stackoverflow.com/a/18242720
+        certStore = XS.makeCertificateStore sc
+        cache :: XV.ValidationCache
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:262]:
```suggestion
```

File: src/Simplex/Messaging/Transport.hs [hunk 22]
```diff
@@ -223,6 +234,52 @@ loadTLSServerParams certificateFile privateKeyFile =
...
+    else pure [XV.UnknownCA]
+  where
+    x509validate :: IO [XV.FailedReason]
+    x509validate = XV.validate X.HashSHA256 hooks checks certStore cache serviceID cc
+      where
+        hooks = XV.defaultHooks
+        checks = XV.defaultChecks {XV.checkLeafV3 = False} -- TODO create v3 certificates? https://stackoverflow.com/a/18242720
+        certStore = XS.makeCertificateStore sc
+        cache = XV.exceptionValidationCache [] -- we don't store certificate Fingerprints on clients
+        serviceID :: XV.ServiceID
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Transport.hs:263]:
```suggestion
```

File: src/Simplex/Messaging/Agent/Protocol.hs [hunk 23]
```diff
@@ -351,11 +351,11 @@ agentMessageP =
...
-smpServerP = SMPServer <$> server <*> optional port <*> optional kHash
-  where
-    server = B.unpack <$> A.takeWhile1 (A.notInClass ":#,; ")
-    port = A.char ':' *> (B.unpack <$> A.takeWhile1 A.isDigit)
-    kHash = C.KeyHash <$> (A.char '#' *> base64P)
+smpServerP = do
+  host <- B.unpack <$> A.takeWhile1 (A.notInClass ":#,; ")
+  port <- optional $ A.char ':' *> (B.unpack <$> A.takeWhile1 A.isDigit)
+  kHash <- C.KeyHash <$> (A.char '#' *> base64P)
+  pure SMPServer {host, port, keyHash = Just kHash}
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent/Protocol.hs:358]:
```suggestion
smpServerP = SMPServer <$> server <*> optional port <*> kHash
  where
    server = B.unpack <$> A.takeWhile1 (A.notInClass ":#,; ")
    port = A.char ':' *> (B.unpack <$> A.takeWhile1 A.isDigit)
    kHash <- (Just . C.KeyHash) <$> (A.char '#' *> base64P)
```

File: src/Simplex/Messaging/Agent/Protocol.hs [hunk 24]
```diff
@@ -351,11 +351,11 @@ agentMessageP =
 
 -- | SMP server location parser.
 smpServerP :: Parser SMPServer
-smpServerP = SMPServer <$> server <*> optional port <*> optional kHash
+smpServerP = SMPServer <$> server <*> optional port <*> kHash
   where
     server = B.unpack <$> A.takeWhile1 (A.notInClass ":#,; ")
     port = A.char ':' *> (B.unpack <$> A.takeWhile1 A.isDigit)
-    kHash = C.KeyHash <$> (A.char '#' *> base64P)
+    kHash <- (Just . C.KeyHash) <$> (A.char '#' *> base64P)
```
Comment (epoberezkin) [inline src/Simplex/Messaging/Agent/Protocol.hs:358]:
```suggestion
    kHash = (Just . C.KeyHash) <$> (A.char '#' *> base64P)
```

**Review by epoberezkin — APPROVED** [review]
